src/coreclr/src/vm/generics.cpp

```diff
@@ -267,13 +267,14 @@ ClassLoader::CreateTypeHandleForNonCanonicalGenericInstantiation(
     // problem however because whenever we need to load a value from the dictionary of this type beyond its size, we
     // will expand the dictionary at that point.
     DWORD cbInstAndDict = pOldMT->GetInstAndDictSize();
+    DWORD cbInstAndDictPlusBackPointer = (cbInstAndDict != 0 ? cbInstAndDict + sizeof(Dictionary *) : 0);
 
     // Allocate from the high frequence heap of the correct domain
     S_SIZE_T allocSize = safe_cbMT;
     allocSize += cbOptional;
     allocSize += cbIMap;
     allocSize += cbPerInst;
-    allocSize += cbInstAndDict;
+    allocSize += cbInstAndDictPlusBackPointer;
```

Comment:
Nit: The starting dictionary does not really need to have the back pointer. It is always going to be NULL. You can tell that it is the starting dictionary by comparing actual dictionary size with GetNumInitialSlots

Reply:
Hmm, methodtablebuilder seems to have its own calculation of the initial dictionary size around

https://github.com/dotnet/runtime/blob/5fd9958c08da7cb65b1a138f65242e4dc67b710f/src/coreclr/src/vm/methodtablebuilder.cpp#L7081

In my local debugging, type dictionaries usually had initial size equal to 0x200; can you please clarify what the check should look like in this case?

I was originally trying to mark the initial directory with zero bit set in the size but for now I rolled it back as an optimization that can be addressed later once the basic algorithmic change is shown to work.

Reply:
OK, I think I finally understand what you meant - using the initial slot count in the DictionaryLayout for the check. Fixed in 3rd commit, thanks for the suggestion.

Reply:
Hmm, I must be still missing something here, apparently it still doesn't work. The code at

https://github.com/dotnet/runtime/blob/06963eaedfe666c087a9fabb89b3f4bca57634c7/src/coreclr/src/vm/genmeth.cpp#L396

apparently allocates an "initial dictionary" with size equal to "current method dictionary layout size" which is different from the "initial slot count" on the dictionary layout. With the back pointers everything seems to be working so I hope this is the last hurdle I need help with.

Reply:
I think this can be solved by moving the accounting for the back-pointer into `GetDictionarySizeFromLayout`:

```
...
    if (pDictLayout != NULL)
    {
        WORD slots = VolatileLoadWithoutBarrier(&pDictLayout->m_numSlots);

        bytes += sizeof(TADDR);                  // Slot for dictionary size
        bytes += slots * sizeof(TADDR);       // Slots for dictionary slots based on a dictionary layout

        if (pDictLayout->m_numInitialSlots < slots)
        {
            bytes += sizeof(PTR_Dictionary);  // Dictionary back-pointer for expanded dictionaries
        }
    }
```

And adding an extra condition for exiting the patching loop on null back-pointers.

Reply:
Sounds great, will do, thank you!

Reply:
Hehe, it's got an interesting wrinkle - we still need two different methods otherwise we'd change the existing semantics of the "size slot"; I have basically kept your idea but I had to introduce a pair of functions, <code>GetDictionarySlotSizeFromLayout</code> vs. <code>GetDictionaryAllocSizeFromLayout</code> where the former is the value of the "size slot" and the latter is the "total number of bytes" including the back pointer.

Reply:
Fixed in 5th commit (using the two methods). It's somewhat uglier than what you originally proposed but it's still relatively well encapsulated I believe.